### PR TITLE
[RISCV][CFI] Emit cfi_offset for every callee-saved vector registers

### LINKED ISSUE #94500: [RISCV] Vector Store/Load whole-register for multiple registers only has DWARF CFI for one register
As seen in our tests (https://github.com/llvm/llvm-project/blob/0295c2ada4d9964a2b6289c905e320bd42461f99/llvm/test/CodeGen/RISCV/rvv-cfi-info.ll#L80C1-L84C8), when spilling vector registers to the stack, we only emit DWARF CFI directives for the register named in the store. Note, that in the code listing in the test, there are other `.cfi_directives`, but those are for other stores earlier in the function.

```asm
 vs4r.v v4, (a0) ;; Unknown-size Folded Spill
;; ...
.cfi_escape 0x10, 0x64, 0x0b, 0x11, 0x60, 0x22, 0x11, 0x78, 0x92, 0xa2, 0x38, 0x00, 0x1e, 0x22 ;; $v4m4 @ cfa - 32 - 8 * vlenb
```

This seems like an oversight in [createDefCFAOffset](https://github.com/llvm/llvm-project/blob/f4681be06b465736cc993b114dd8e2625d37b779/llvm/lib/Target/RISCV/RISCVFrameLowering.cpp#L500), or maybe it's callers.

It may be worth considering extending the DWARF registers with some kind of pseudo register for these groups, to keep the debug info compact.

### PR DESCRIPTION
**Pull request by 4vtomat**
The grouped vector register is modeled as a single register, e.g. V2M2,
which is actually V2 and V3. We need to decompose the grouped vector
register(if any) to individual vector register when emitting CFIs in
prologue.

Fixed https://github.com/llvm/llvm-project/issues/94500

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 1]
```diff
@@ -53,6 +53,35 @@ static const std::pair<MCPhysReg, int8_t> FixedCSRFIMap[] = {
...
+// V8   -> {RISCV::V8, 1}
+static std::pair<MCPhysReg, int8_t> getCSBaseVRegLMULPair(MCPhysReg VR) {
+  assert(((VR >= RISCV::V1 && VR <= RISCV::V7) ||
+          (VR >= RISCV::V24 && VR <= RISCV::V31) ||
+          (VR >= RISCV::V2M2 && VR <= RISCV::V6M2) ||
+          (VR >= RISCV::V24M2 && VR <= RISCV::V30M2)) &&
+         "Invalid VR");
+
+  static constexpr std::pair<MCPhysReg, int8_t> VRegLMULLUT[] = {
+      {/*V2M2*/ RISCV::V2, 2},   {/*V4M2*/ RISCV::V4, 2},
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:68]: This needs static_asserts to make sure the registers are really in this order.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -1554,12 +1583,15 @@ void RISCVFrameLowering::emitCalleeSavedRVVPrologCFI(
     // Insert the spill to the stack frame.
     int FI = CS.getFrameIdx();
     if (FI >= 0 && MFI.getStackID(FI) == TargetStackID::ScalableVector) {
-      unsigned CFIIndex = MF->addFrameInst(
-          createDefCFAOffset(*STI.getRegisterInfo(), CS.getReg(), -FixedSize,
-                             MFI.getObjectOffset(FI) / 8));
-      BuildMI(MBB, MI, DL, TII.get(TargetOpcode::CFI_INSTRUCTION))
-          .addCFIIndex(CFIIndex)
-          .setMIFlag(MachineInstr::FrameSetup);
+      auto VRegLMULPair = getCSBaseVRegLMULPair(CS.getReg());
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1586]: Can we use TRI->getSubReg with `RISCV::sub_vrm1_0` find the starting register?

Can we get the LMUL by checking calling `contains` on the 4 possible vector register classes?

Comment (4vtomat) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1586]: Yes, good idea, thanks!

**Comment by ilovepi** [comment]
LGTM from me. I'll let @topperc make the final call, but I like this approach much better. Also, thanks for fixing this up!

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**